hud-scripts/update_pending.py

```diff
@@ -0,0 +1,129 @@
+            return json.load(data)
+    except HTTPError as err:
+        if err.code == 403 and all(key in err.headers for key in ['X-RateLimit-Limit', 'X-RateLimit-Used']):
+            print(f"Rate limit exceeded: {err.headers['X-RateLimit-Used']}/{err.headers['X-RateLimit-Limit']}")
+        raise
+
+
+def get_circleci_token() -> str:
+    token_file_path = os.path.join(os.getenv('HOME'), '.circleci_token')
+    token = os.getenv('CIRCLECI_TOKEN')
+    if token is not None:
+        return token
+    if not os.path.exists(token_file_path):
+        return None
+    with open(token_file_path) as f:
+        return f.read().strip()
+
+
+def circleci_fetch_json(url: str) -> Union[Dict[str, Any], List[Dict[str, Any]]]:
+    token = get_circleci_token()
+    headers = {'Accept': 'application/json'}
+    if token is not None:
+        headers['Circle-Token'] = token
+    with urlopen(Request(url, headers=headers)) as data:
+        return json.load(data)
+
+
+def gh_fetch_multipage_json(url: str, params: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
+    if params is None:
+        params = {}
+    assert "page" not in params
+    page_idx, rc, prev_len, params = 1, [], -1, params.copy()
+    while len(rc) > prev_len:
+        prev_len = len(rc)
+        params["page"] = page_idx
+        page_idx += 1
+        rc += gh_fetch_json(url, params)
+    return rc
+
+
+def gh_get_ref_statuses(org: str, project: str, ref: str) -> Dict[str, Any]:
+    url = f'https://api.github.com/repos/{org}/{project}/commits/{ref}/status'
+    params = {"page": 1, "per_page": 100}
+    nrc = rc = gh_fetch_json(url, params)
+    while "statuses" in nrc and len(nrc["statuses"]) == 100:
+        params["page"] += 1
+        nrc = gh_fetch_json(url, params)
+        if "statuses" in nrc:
+            rc["statuses"] += nrc["statuses"]
+    return rc
+
+
+def map_circle_status(status: str) -> str:
+    if status == "running":
+        return "pending"
+    return status
+
+
+def update_pending(branch: str = "master") -> None:
+    commit_index = s3_get_json(bucket_name, f'{branch}/index.json', [])
+    for idx, item in enumerate(commit_index):
+        commit_id = item['id']
+        title = item['message'].split("\n")[0]
+        timestamp = datetime.fromisoformat(item['timestamp']).replace(tzinfo=None)
+        if datetime.utcnow() - timestamp < timedelta(hours=5):
+            print(f"[{idx}/{len(commit_index)}] {title} ( {commit_id} ) is skipped as it was merged less than 5 hours ago")
+            continue
+        has_pending, has_updates = False, False
+        job_statuses = s3_get_json(bucket_name, f'{branch}/{commit_id}.json', {})
+        for (name, value) in job_statuses.items():
+            status = value['status']
+            build_url = value['build_url']
+            if status not in ['success', 'skipped', 'error', 'failure']:
+                rc = re.match("https://circleci.com/gh/pytorch/pytorch/(\\d+)\\?", build_url)
```

Comment:
I saw similar things for github action as well, e.g. 

```
    "Lint / flake8-py3": {
        "build_url": "https://github.com/pytorch/pytorch/runs/2651879076",
        "status": null
    },
```

With 
```
curl 'https://s3.amazonaws.com/ossci-job-status/master/a7fdd487e50d99976dd018a2e0edc7d49e9fb968.json' 
```

However, the status of https://github.com/pytorch/pytorch/runs/2651879076 is green.

Reply:
Oh, actually, the event is indeed with status `null`, the event looks like from the auto cancel workflow, so maybe we don't need to take care of it for now, though we should probably just display "skipped" instead of "?" on the HUD.

```
{
    "commitId": "a7fdd487e50d99976dd018a2e0edc7d49e9fb968",
    "build_url": "https://github.com/pytorch/pytorch/runs/2651879076",
    "job_name": "Lint / flake8-py3",
    "status": null,
    "committer": "facebook-github-bot",
    "commit_source": "master",
    "bucket": "ossci-job-status"
}
```

![image](https://user-images.githubusercontent.com/658840/119384472-9f067f80-bc79-11eb-84c7-55cddec12ffc.png)

Reply:
We definitely do, this only takes care of CircleCI statuses.

Reply:
given the complexity and maintenance cost of fixing both circleci and GHA and future sync issue, do you think we can go by adding a redis lock to make the event handler sequential? It may be straightforward to just add a few lines in the original lambda handler function 

```
with redis.lock:
  # do the original thing
```

Reply:
Updated to handle GHR statuses.

Goal of this script is not to hide race conditions in the lambda, but rather populate database in case notifications are missing  and also pre-populate it for new repos/branches.

And please do not hesitate submitting PR that fixes race with redis.lock

Reply:
thanks, and yes, we do need this script to fix previous missing mismatch status and potentially future new repos/branches.